### ISSUE TITLE
http_get() is super slow

### ISSUE THREAD

**Issue by retronerdboy** [SuperIlu/DOjS#15]
It takes ~40 seconds to load 10kb of data while wget (which also uses WATTCP) reaches download speed of ~100kb/s.

I've tested this in FreeDOS 1.2.

**Comment by SuperIlu** [comment]
Which environment did you use?

**Comment by retronerdboy** [comment]
I use an old computer to run it so that's the reason probably, will wait for the native version :-)

**Comment by SuperIlu** [comment]
If you have any skills with Javascript and HTTP you can have a look into jsboot/socket.js and try to improve the HTTP-body reading part. This should be the culprit ;)

I make no promises on when the native implementation will come out...

**Comment by retronerdboy** [comment]
I've an idea how to improve it but so far I've problems compiling the project on the watt32 step, I get this error:

```
echo 'const char *w32_cflags = "-Os -g -I. -I../inc -W -Wall -Wno-strict-aliasing -DWATT32_BUILD ";' > build/djgpp/cflags.h
/bin/sh: 1: cannot create build/djgpp/cflags.h: Directory nonexistent
```

I've created the `build/djgpp` directory but then I get different errors:

```
In file included from ../inc/sys/socket.h:51:0,
                 from socket.h:37,
                 from accept.c:51:
../inc/sys/werrno.h:89:12: fatal error: sys/djgpp.err: No such file or directory
   #include <sys/djgpp.err>
            ^~~~~~~~~~~~~~~
compilation terminated.
DJGPP.MAK:171: recipe for target 'build/djgpp/accept.o' failed
```

Maybe there is some step missing in the compilation instructions as watt32 was introduced not so long ago?

**Comment by SuperIlu** [comment]
Ok, several questions:

- I guess you followed the build-env setup instructions from README.md?
- Did you download an archive or did you clone the repo?
- What do you want to achieve by compiling DOjS.EXE yourself? If you just wanna improve http_get() you can simply edit jsboot/socket.js. No need to compile.

You are right that the build-instructions haven't been updated fo quite a while, but as far as I remember nothing in the environment has changed. I regularly build DOjS on two different machines.

Can you maybe provide me with the full output of "make" for both cases. Or just with the output of "make distclean libwatt32"?

**Comment by SuperIlu** [comment]
I just pushed a small update which should resolve the build problem.
I tried it with a clean checkout to a case sensitive file system on my WSL and it all compiled...

**Comment by retronerdboy** [comment]
Thanks! I got it compiled, although I had to change one more thing: `DJGPP` path in `glide3x/texus/Makefile` (similar to Makefile in root directory as described in readme)

I had to compile the project because I wanted to access the `Socket_ReadBytes` from `socket.c`, currently it looks like this:

```
PROTDEF(J, Socket_ReadByte, TAG_SOCKET, "ReadByte", 0);
PROTDEF(J, Socket_ReadBytes, TAG_SOCKET, "ReadByte", 0);
```

I thought maybe there's some overriding going on (if no arg provided than use `Socket_ReadByte`, otherwise `Socket_ReadBytes`) but it always seemed to use the `Socket_ReadByte` function.

Anyway, I changed the data reading in `http_get` to use `ReadBytes` with the amount returned by `DataReady` and it works a lot faster, but still too slow :-) I want to move the whole data reading part to a separate function in `socket.c` and see if that helps.

**Comment by SuperIlu** [comment]
You found a typo. Damn, I should have simplified that macro long ago!
And yes, using ReadBytes was my first idea, too.

Btw. arrays are quite slow with MuJS because internally they convert the int index to a string and store the data as an object property :(

**Comment by SuperIlu** [comment]
The next release will include a dedicated class for dealing with file/socket/zip IO operations.
This seems to speed up things a lot when dealing with binary data ("Bytes" are JS arrays, "Ints" is the new IntArray class):
```
Runtime for 'File.ReadBytes() := 88452' 	is  1s 790ms
Runtime for 'File.ReadInts() := 88452'  	is      60ms
Runtime for 'File.WriteBytes()' 		is     470ms
Runtime for 'File.WriteInts()' 			is      40ms
Runtime for 'Zip.WriteBytes()' 			is     670ms
Runtime for 'Zip.WriteInts()' 			is     230ms
Runtime for 'Array.Push() := 20000' 		is     700ms
Runtime for 'IntArray.Push() := 20000' 		is     290ms
Runtime for 'Array.Get() == 199990000' 		is     300ms
Runtime for 'IntArray.Get() == 199990000' 	is     280ms
Runtime for 'Array.Pop() == 1994950' 		is       0ms
Runtime for 'IntArray.Pop() == 1994950' 	is      10ms
Runtime for 'Array.Shift() == 4950' 		is 15s 740ms
Runtime for 'IntArray.Shift() == 4950' 		is      50ms
```

**Comment by SuperIlu** [comment]
Good news: I have added loadable cURL support (including TLS) to DOjS. I will release that in the next week if everything is smooth...

**Comment by SuperIlu** [comment]
cURL and dedicated binary data class implemented in release V1.4.0
